app/cdap/components/PipelineDiff/store/pipelineUtil.ts

```diff
@@ -34,6 +44,72 @@ export function getReactflowPipelineGraph(config) {
   return { nodes, connections };
 }
 
-export function computePipelineDiff(pipeline1, pipeline2) {
-  return [];
+function preprocessPipeline(pipeline: IPipelineConfig) {
+  const stageMap: IStageMap = {};
+  const connectionMap: IConnectionMap = {};
+  const stageNameToPluginName = {};
+
+  for (const stage of pipeline.stages) {
+    const stageMapKey = `${stage.plugin.name}__${stage.name}`;
+    stageMap[stageMapKey] = stage;
+    stageNameToPluginName[stage.name] = stage.plugin.name;
+  }
+
+  for (const connection of pipeline.connections) {
+    const fromName = `${stageNameToPluginName[connection.from]}__${connection.from}`;
+    const toName = `${stageNameToPluginName[connection.to]}__${connection.to}`;
+    // TODO: I am not sure if the users are allowed to use % in their names, other character
+    // that is guarenteed to be not used is preferred
+    const connectionMapKey = `${fromName}%${toName}`;
```

Comment:
% is allowed in names. So actually it's not safe to even split on it. May need to rethink about this